### ISSUE TITLE
Tasks for collecting/submitting field data to fulfill sponsor's mission agenda

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#898]
**Is your feature request related to a problem? Please describe.**
- It's known in mars-sim that each space agency or sponsor has a particular mission agenda. 

**Describe the solution you desire to see**
- Create a new Collect Field Data `Task` and `MetaTask`
- Create a data score and record what kind of data and how much time 

**Describe alternatives you've considered**
- Create a `Mission` that dedicate for meeting the expectation of the sponsor 
 
**Describe the rationale**
- Collecting vital field data and submitting them to his/her superior in Earth's mission control is the major way in fulfilling the sponsoring agency's agenda. 
- This is because, other than the survival of the settlers, the job of a commander must also appease the sponsor enough to ensure proper support be provided and resupply missions be lined up in foreseeable future. 

**Specifications  (please complete the following information):**
 - pre 3.6.0 build 8102

**Additional context**
- Q: how do we improve the usefulness of the `ReportMissionControl` task in conjunction with adding this new task at the same time ?
- In future, we may also add in what ways and how earth mission controls may ask settlement commanders to do that will develop/demonstrate the 3 capabilites as defined in mission agenda.

**Comment by mokun** [comment]
@bevans2000 

While we were talking about a sponsor's relationship with a country, one may ask, what exactly is the role of a sponsor in a settlement ? 

We haven't done much to expound what a sponsor can do for a settlement. 

We do have modifiers from sponsor's agenda but they need to be spelled out clearly that they are sponsor's expectation.

There should be differences between what a sponsor wants, what the settlement leadership wants and even what the majority of settlers in a settlement wants.

Do you agree and what UI elements should we create to represent this type of expectations and conflicting objectives ?

**Comment by bevans2000** [comment]
I think the Agenda of the Authority can influence something but I can't remember what. Might be science.

**Comment by Urwumpe** [comment]
> While we were talking about a sponsor's relationship with a country, one may ask, what exactly is the role of a sponsor in a settlement ?

IMHO he is to provide two things:

- Provide settlements with supplies, research goals and settlers until there is autarky in all that aspects.
- Provide the player/supervisor/god with some challenges to become creative and interact with the simulation.
> 
> We haven't done much to expound what a sponsor can do for a settlement.

There are some possibilities, as mentioned above. For example, a sponsor could have patents for technologies, that others don't, like a special building, a different rover chassis, lighter EVA suits, medicine. Maybe we could also include the mouse company for entertainment, like kickstarting a martian movies business.

> 
> We do have modifiers from sponsor's agenda but they need to be spelled out clearly that they are sponsor's expectation.

Yes. Maybe we should start with the basics, like "produce x of type" or "build x of building" or "provide quarters of x settlers" or "increase production of x to y"
> 
> There should be differences between what a sponsor wants, what the settlement leadership wants and even what the majority of settlers in a settlement wants.

Yes. Or what the individual wants. Personal goals could also be a driving force.

> 
> Do you agree and what UI elements should we create to represent this type of expectations and conflicting objectives ?

Well, we should separate expectations and clear goals. Goals are quantifyable and testable, e.g. checked if the conditions are met. Expectations on the other hand can be rather undefined, change quickly, even for sponsors, and have changing priorities. But can be used as input for creating goals. Maybe we defined a WIP limit for the number of goals an entity can have, like 3 goals for sponsors and settlement leadership and one personal goal for settlers. When one is fulfilled, we could for the start create a new goal based on the highest priority expectation and the largest deficit between current state and expectation.

For goals, we could make a special type of panel, that we could stack vertically and where all important information is displayed in a standardized format. It should have a Ícon for the type of entity that sets the goal, and there should be a listing of all major goals as UI view. Maybe with filtering.

Expectations might be more interesting if they are not easy to see, expect for debugging and testing. Maybe as quotes or thoughts of the settlers, or a press release of a sponsor or statement of an official of that sponsor. 

We once had a feature tested called Mars Net, maybe we can revive this with some stricter design and more clear design. Might be a good case for a project within mars-sim to return to the idea if you agree.

**Comment by mokun** [comment]
> a sponsor could have patents for technologies, that others don't, like a special building, a different rover chassis, lighter EVA suits, medicine...

I like that ! Since it's a patent, other settlements can potentially license that...

**Comment by mokun** [comment]
> Well, we should separate expectations and clear goals. Goals are quantifyable and testable, e.g. checked if the conditions are met. 

Yes. A goal needs to be very specific.

> Expectations on the other hand can be rather undefined, change quickly, even for sponsors, and have changing priorities. But can be used as input for creating goals.

I recalled playing Sims 3 a long time ago. Every person has a desire/goal at a given moment. Fulfilling it earns you a bonus on something... 

Of course, after a while, even that becomes methodical and monotonous is trying to make things happen for a person or a family... may be the simulation is still simulation and nothing come close to simulating reality yet...

**Comment by Urwumpe** [comment]
> Of course, after a while, even that becomes methodical and monotonous is trying to make things happen for a person or a family... may be the simulation is still simulation and nothing come close to simulating reality yet...

That is why I hate knowing too much about the inner workings of a settlers mind. Its the same with Dwarf Fortress, where at least the bulk of people prevents you optimizing too much.

I would prefer to have settlers, who one day pick a random book from the electronic library, get inspired from it and changes its life to become a master botanist by planting potatoes on Martian soil.

**Comment by mokun** [comment]
I'm going to create a FieldData class to begin simulating the need for data collection in the field.

I'll have a DataManager that will run on the main thread. 

This way, these field data objects can be worked on by anyone from any settlements and even change hand in future.

At the moment, I can think of defining enum DataType as follows: 
1. Ground data
2. Aerial data
3. Ground photo
4. Aerial photo

Scientists and engineers will spend time and resources for processing and refining these field data to increase its quality and integrity. 

At the end, it can be used for writing research papers, to be sold to earn credits and/or to fulfill sponsor's mission requirement.

Anything else you can think of ?

**Comment by bevans2000** [comment]
Ok but would this data be normally shared between Settlements?
Why not extend the Surface features as that already manages places of interest which are MineralSites

**Comment by mokun** [comment]
> Ok but would this data be normally shared between Settlements? 

Yes it will be. Since our scientific studies already span across settlers from different time zones, right ?

> Why not extend the Surface features as that already manages places of interest which are MineralSites

Do you mean SurfacePOI ? But it's an interface.

<img width="797" height="566" alt="Image" src="https://github.com/user-attachments/assets/0af19f55-7dae-404c-bea0-da6bca8e35e6" />

MineralSite and MineralDeposit are definitely mineral-focused.

I see field data can go beyond just the mineral data.

Essentially there are weather data, mineral data, areological data, and whatever data you can name.

So when it comes to the mineral, I'll have MineralData that will extend FieldData, etc.

**Comment by bevans2000** [comment]
But still data about a point on the Surface? I'll leave it to you but SurfacePOI comes with some enablers like a ready to go map later. So you might want to show any field data in a geographical view. You might want to parent data items from the same spot against a single  parent

**Comment by mokun** [comment]
> So you might want to show any field data in a geographical view. You might want to parent data items from the same spot against a single parent

Right. I would love that.

I have made FieldData implementing SurfacePOI.

SurfacePOI currently doesn't have anything particularly useful but it could be developed further to host more useful features later.

**Comment by mokun** [comment]
In terms of creating a new task for gathering the field data, I plan to

1. Create GatherDataMeta (similar to DigLocalMeta) that implements SettlementMetaTask - this avoids having to call assignTask() in a mission and causes walking bug
2. Create GatherData (similar to DigLocal)
3. Create GatherGroundData and GatherAerialData

In terms of creating GatherGroundData,

1. Have a few members assemble equipment/instruments on site
2a. Wait for enough field data to be captured
2b. Have a few members walk around and take photos
3. Have a few members Disassemble equipment/instrument and return to the vehicle

> Note: 2a and 2b are executed in parallel.

In terms of creating GatherAerialData,

1. Create a MonitorMission
2. Dispatch drones for settlement inspection / mission monitoring / weather monitoring
3. Fly into the destination
4. Capture aerial data or ground data
5. Return to the settlement

**Comment by bevans2000** [comment]
Sounds good. Try to put all the selection logic of people and vehicles into the MetaTask and not the Mission. So the mission just executes the objective. It should be easy to add it to the mission wizard so you can create them manually as well. The Exploration mission is a good example of the new cleaner approach.

**Comment by mokun** [comment]
> The Exploration mission is a good example of the new cleaner approach.

How cleaner is this approach ? 

It's still using assignTask() to execute ExploreSite task. There's still a risk for that person to get stuck and die.

Therefore, I'd like to create ExploreSiteMeta to trigger ExploreSite only and not assignTask since assignTask bypass the CacheCreator in TaskManager and a person should not be forced to do one thing and should be given a probability of task to choose from. 

As a matter of fact, I'd like to create VehicleMetaTask, instead of SettlementMetaTask

So ExploreSiteMeta, GatherDataMeta, etc. will implement VehicleMetaTask, if they start from a vehicle.

In Exploration, 
```java
/**
	 * Updates the explored site and start an ExploreSite Task.
	 * 
	 * @param person
	 */
	@Override
	protected boolean performEVA(Person person) {
		
		boolean canAssign = false;
		
		// If person can explore the site, start that task.
		if (ExploreSite.canExploreSite(person)) {
			
			// Add new explored site if just starting exploring.
			if (currentSite == null) {
				currentSite = retrieveASiteToClaim();
				if (currentSite == null) {
					abortMission(INVALID_EXPLORATION_SITE);
					return false;
				}
			}
			canAssign = assignTask(person, new ExploreSite(person, currentSite, getRover(), this));
			
			if (canAssign) {
				logger.info(person, 20_000, "Ready to explore site and collect rocks.");
				
				// Update exploration site completion.
				double timePassed = getPhaseTimeElapsed();
				double completion = timePassed / STANDARD_TIME_PER_SITE;
				if (completion > 1D) {
					completion = 1D;
				}
				else if (completion < 0D) {
					completion = 0D;
				}		
				
				fireMissionUpdate(SITE_EXPLORATION_EVENT, getCurrentNavpointDescription());

				objective.updateSiteCompletion(getCurrentNavpointDescription(), completion);
			}
		}
		else {
			endEVATasks();
		}

		return true;
	}
```

**Comment by mokun** [comment]
@bevans2000,

I'm planning to use these 5 instruments in GatherData,

```java
	private Set<Integer> dataInstrumentSet = Set.of(
			ItemResourceUtil.GAS_CHROMATOGRAPH_ID, 
			ItemResourceUtil.IR_SPECTROMETER_ID,
			ItemResourceUtil.MASS_SPECTROMETER_ID, 
			ItemResourceUtil.XRAY_SPECTROMETER_ID, 
			ItemResourceUtil.HEAT_PROBE_ID);
			
	private Map<Integer, Person> operatorMap = new HashMap<>();
```

They are all Item Resources or parts. They should be considered portable lab equipment and therefore should be of Equipment class in near future.

Currently, we have a few GenericContainer and EVASuit. Their EquipmentType enum are as follows : 

```java
**
 * The EquipmentType enum class is used for distinguishing between various type of equipments
 */
public enum EquipmentType implements Named {
	
	// non-container
	EVA_SUIT, 

	// Container 
	BAG, BARREL, GAS_CANISTER, LARGE_BAG, SPECIMEN_BOX,
	THERMAL_BOTTLE, WHEELBARROW;
```

Q1: Should we have just one instrument type for the 3 spectrometers and the gas chromatograph or 4 separate equipment type ?

Q2: how do you like to handle adding these instruments ? 

Note that the heat probe should remain as a simple part.

**Comment by bevans2000** [comment]
It really comes down to what level we want to simulate them. Equipment are either dynamic entities like EVAsuit or can hold other resources. They are more heavyweight than Item resources. 
Personally I think how we model Equipment is not 100% correct at the moment. Container feels like the design is wrong but it works so can live with it.
For these  instruments I would keep them as Item resource for now. It will allow you to complete your mission work without getting too complex.

**Comment by mokun** [comment]
> Personally I think how we model Equipment is not 100% correct at the moment. Container feels like the design is wrong but it works so can live with it.

Exactly. Since each container is only a amount resource holder, there are other ways to emulate what it does with less expensive implementation. This is what I have wanted to define Bin to achieve.

> equipment are either dynamic entities like EVAsuit

Yes, as for instruments, they will be like EVASuit that make impact in the environment to produce result. 

So it feels like being an object is more appropriate and will be better in the long run.

As we know Item Resource or Part doesn't record any individual history. Two same parts are indistinguishable from each other. 

Therefore, sooner or later, we will need to make each instrument a equipment.
 
>  It will allow you to complete your mission work without getting too complex.

I could do a quick implementation of an Instrument class that extends Equipment.

This way, we won't have to waste time in future converting them from Part to Equipment.

**Comment by mokun** [comment]
Actually I'm going to first create DataRecorder that extends Equipment class. 

It will be used as a container for holding FieldData objects.

For now, the spectrometers will remain as just parts.

**Comment by bevans2000** [comment]
Ok but make sure you create unit tests

**Comment by mokun** [comment]
> Ok but make sure you create unit tests

Yes I created a DataRecorderTest.

Also, I can't find any unit test on the EVASuit class. 

EVAOperationTest did technically tested the ability to do EVA and add O2 and H2O to an EVASuit but it didn't proceed to test other aspects there. 

So I created EVASuitTest as well.

**Comment by mokun** [comment]
btw, I did a commit on the basic setup. Still working on getting them ready for prime time.

**Comment by mokun** [comment]
After the commit yesterday, settlers are able to generate ground data when using instruments to look for water ice during the data collection phase of GatherData task. 

But I still need to tie that to how it will impact digging ice and connect more dots.

More will follow.

<img width="987" height="731" alt="Image" src="https://github.com/user-attachments/assets/1c39cb61-15b1-4983-982b-7f8a4fc8145c" />
